// @stellar/design-system/build/components/Layout/index.js
import { __assign } from "tslib";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useEffect } from "react";
import { Icon } from "../../icons";
import { Logo } from "../../logos";
import { ProjectLogo } from "../ProjectLogo";
import { Link } from "../Link";
import { ThemeMode, ThemeSwitch } from "../ThemeSwitch";
import { NavButton } from "../NavButton";
import "./styles.scss";
export var Layout = function () {
};
export var Inset = function (_a) {
    var children = _a.children;
    return (_jsx("div", { className: "Layout__inset", children: children }));
};
export var Content = function (_a) {
    var children = _a.children;
    return (_jsx("div", { className: "Layout__content", children: children }));
};
var stringToCamelcase = function (str) {
    return str
        .replace(/(?:^\w|[A-Z]|\b\w)/g, function (word, index) {
        return index === 0 ? word.toLowerCase() : word.toUpperCase();
    })
        .replace(/\s+/g, "");
};
export var Header = function (_a) {
    var projectId = _a.projectId, projectTitle = _a.projectTitle, projectLink = _a.projectLink, hasThemeSwitch = _a.hasThemeSwitch, onThemeSwitchActionEnd = _a.onThemeSwitchActionEnd, onSignOut = _a.onSignOut, showButtonBorder = _a.showButtonBorder, menu = _a.menu, contentCenter = _a.contentCenter, contentRight = _a.contentRight, disableSetThemeOnLoad = _a.disableSetThemeOnLoad;
    useEffect(function () {
        if (disableSetThemeOnLoad) {
            return;
        }
        if (!hasThemeSwitch) {
            document.body.classList.add(ThemeMode.LIGHT);
        }
    }, [disableSetThemeOnLoad, hasThemeSwitch]);
    return (_jsx("div", { className: "Layout__header", children: _jsxs(Layout.Inset, { children: [_jsx("div", { className: "Layout__header--left", children: _jsx(ProjectLogo, { title: projectTitle, link: projectLink }) }), contentCenter ? (_jsx("div", { className: "Layout__header--center", children: contentCenter })) : null, _jsxs("div", { className: "Layout__header--right", children: [contentRight !== null && contentRight !== void 0 ? contentRight : null, onSignOut ? (_jsx(Link, { id: "sign-out-button", role: "button", onClick: onSignOut, children: "Sign out" })) : null, hasThemeSwitch ? (_jsx(ThemeSwitch, { storageKeyId: "stellarTheme:".concat(stringToCamelcase(projectId)), onActionEnd: onThemeSwitchActionEnd, disableSetThemeOnLoad: disableSetThemeOnLoad })) : null, (menu === null || menu === void 0 ? void 0 : menu.isEnabled) ? (_jsx(NavButton, { id: "open-side-nav-button", title: "Open side navigation", onClick: menu.onOpen, icon: _jsx(Icon.Menu01, {}), showBorder: showButtonBorder })) : null] })] }) }));
};
export var Footer = function (_a) {
    var marginTop = _a.marginTop, hideLegalLinks = _a.hideLegalLinks, hideTopBorder = _a.hideTopBorder, gitHubLink = _a.gitHubLink, gitHubLabel = _a.gitHubLabel, children = _a.children;
    var customStyle = __assign(__assign({}, (marginTop ? { "--Layout-footer-margin-top": marginTop } : {})), (hideTopBorder ? { "--Layout-footer-border-color": "transparent" } : {}));
    return (_jsx("div", { className: "Layout__footer", style: customStyle, children: _jsx(Inset, { children: _jsxs("div", { className: "Layout__footer__content", children: [hideLegalLinks ? null : (_jsxs("div", { className: "Layout__footer__legal", children: [_jsx(Link, { href: "https://www.stellar.org/terms-of-service", variant: "secondary", children: "Terms of Service" }), _jsx(Link, { href: "https://www.stellar.org/privacy-policy", variant: "secondary", children: "Privacy Policy" })] })), _jsxs("div", { className: "Layout__rightSide--inline", children: [gitHubLink ? (_jsx(Link, { href: gitHubLink, variant: "secondary", icon: _jsx(Logo.Github, {}), iconPosition: "left", children: gitHubLabel !== null && gitHubLabel !== void 0 ? gitHubLabel : "GitHub" })) : null, children] })] }) }) }));
};
Layout.Inset = Inset;
Layout.Inset.displayName = "Layout.Inset";
Layout.Content = Content;
Layout.Content.displayName = "Layout.Content";
Layout.Header = Header;
Layout.Header.displayName = "Layout.Header";
Layout.Footer = Footer;
Layout.Footer.displayName = "Layout.Footer";
