// @stellar/design-system/build/logos.js
import { ReactComponent as Albedo } from "./assets/logos/albedo.svg";
import { ReactComponent as AppStore } from "./assets/logos/app-store.svg";
import { ReactComponent as Apple } from "./assets/logos/apple.svg";
import { ReactComponent as Discord } from "./assets/logos/discord.svg";
import { ReactComponent as Framer } from "./assets/logos/framer.svg";
import { ReactComponent as Freighter } from "./assets/logos/freighter.svg";
import { ReactComponent as Github } from "./assets/logos/github.svg";
import { ReactComponent as GooglePlay } from "./assets/logos/google-play.svg";
import { ReactComponent as Ledger } from "./assets/logos/ledger.svg";
import { ReactComponent as Linkedin } from "./assets/logos/linkedin.svg";
import { ReactComponent as Medium } from "./assets/logos/medium.svg";
import { ReactComponent as Reddit } from "./assets/logos/reddit.svg";
import { ReactComponent as StellarShort } from "./assets/logos/stellar-short.svg";
import { ReactComponent as Stellar } from "./assets/logos/stellar.svg";
import { ReactComponent as Trezor } from "./assets/logos/trezor.svg";
import { ReactComponent as XTwitter } from "./assets/logos/x-twitter.svg";
import { ReactComponent as Youtube } from "./assets/logos/youtube.svg";
export var Logo = {
    Albedo: Albedo,
    AppStore: AppStore,
    Apple: Apple,
    Discord: Discord,
    Framer: Framer,
    Freighter: Freighter,
    Github: Github,
    GooglePlay: GooglePlay,
    Ledger: Ledger,
    Linkedin: Linkedin,
    Medium: Medium,
    Reddit: Reddit,
    StellarShort: StellarShort,
    Stellar: Stellar,
    Trezor: Trezor,
    XTwitter: XTwitter,
    Youtube: Youtube,
};
